pkg/placement/binding-organizer.go

```diff
@@ -20,160 +20,222 @@ import (
+var factorClusterWhatWhereFullKey = Factoring[ClusterWhatWhereFullKey, NonNamespacedDistributionTuple, string /* ep name */]{
+	First: func(cfk ClusterWhatWhereFullKey) Pair[NonNamespacedDistributionTuple, string /* ep name */] {
+		return Pair[NonNamespacedDistributionTuple, string /* ep name */]{
+			First: NonNamespacedDistributionTuple{
+				First: ProjectionModeKey{
+					GroupResource: cfk.Second.First,
+					Destination:   cfk.Third,
+				},
+				Second: ExternalName{
+					Cluster: cfk.First.Cluster,
+					Name:    cfk.Second.Second,
+				},
+			},
+			Second: cfk.First.Name}
+	},
+	Second: func(decomp Pair[NonNamespacedDistributionTuple, string /*ep name*/]) ClusterWhatWhereFullKey {
+		return ClusterWhatWhereFullKey{
+			First:  ExternalName{decomp.First.Second.Cluster, decomp.Second},
+			Second: Pair[metav1.GroupResource, string]{decomp.First.First.GroupResource, decomp.First.Second.Name},
+			Third:  decomp.First.First.Destination,
 		}
-		return ans
-	}
+	},
 }
 
 // simpleBindingOrganizer is the top-level data structure of the organizer.
 // In the locking order it precedes its discovery and its projectionMapProvider,
 // which in turn precedes each projectionPerClusterImpl.
+// A BindingOrganizer is given the change stream of the map
+// from triple(ExternalName, WorkloadPartID, SinglePlacement) to WorkloadPartDetails,
+// and converts that into the changes streams to the WorkloadProjector.
+//
+// For the cluster-scoped objects it proceeds as follows.
+// First there is a collect+project to squeeze out the EdgePlacement object name
+// (leaving its cluster); that reduced map change stream is used for two things.
+// Its key set change stream is extracted and drives the NonNamespacedDistributions.
+// Second, it is grouped by everything but the WorkloadPartDetails and those inner
+// sets of WorkloadPartDetails are run through version conflict resolution; this
+// produces the change stream given to the NonNamespacedModes.
+// (That is less than the full design, which calls for using API discovery to inform
+// a more complex conflict resolution procedure.)
+//
+// Not implemented yet: namespaced objects.
 type simpleBindingOrganizer struct {
-	logger                    klog.Logger
-	discovery                 APIMapProvider
-	resourceModes             ResourceModes
-	eventHandler              EventHandler
-	projectionMapProvider     TransformingRelayMap[ProjectionKey, *projectionPerClusterImpl, *ProjectionPerCluster]
-	projectionMappingReceiver ProjectionMappingReceiver
+	logger        klog.Logger
+	discovery     APIMapProvider
+	resourceModes ResourceModes
+	eventHandler  EventHandler
+	// projectionMapProvider TransformingRelayMap[ProjectionKey, *projectionPerClusterImpl, *ProjectionPerCluster]
+	workloadProjector    WorkloadProjector
+	clusterWhatWhereFull MappingReceiver[ClusterWhatWhereFullKey, ProjectionModeVal]
 	sync.Mutex
-}
+	workloadProjectionSections WorkloadProjectionSections // non-nil only during a transaction!
```

Comment:
Question: Does this inline comment refer to line 195 (`sbo.workloadProjectionSections = WorkloadProjectionSections{}`)?

Reply:
yes

Reply:
Is `WorkloadProjectionSections{}` itself nil when it has several nil fields? I'm not sure.

Reply:
Sorry, typo, the comment should say non-zero rather than non-nil

Reply:
Fixed in latest revision.